File: find-nearest-right-node-in-binary-tree/find-nearest-right-node-in-binary-tree.py
from collections import deque
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findNearestRightNode(self, root: TreeNode, u: TreeNode) -> TreeNode:
        
        q = deque([[root, 0]])
        levels = []
        
        while(q):
            node, level = q.popleft()
            if level >= len(levels):
                levels.append([])
            levels[level].append(node)
            for x in [node.left, node.right]:
                if x:
                    q.append([x, level+1])
​
        for level in levels:
            for i in range(len(level)):
                if level[i] == u:
                    if i < len(level)-1:
                        return level[i+1]
                    else:
                        return None
        
        
        

File: find-nearest-right-node-in-binary-tree/find-nearest-right-node-in-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func findNearestRightNode(root *TreeNode, u *TreeNode) *TreeNode {
    q := make([]*TreeNode, 0)
    q = append(q, root)
    
    for len(q) != 0 {
        level_len := len(q)
        
        for i := range q {
            node := q[0]
            q = q[1:]
            
            if node == u {
                if i != level_len - 1 {
                    return q[0]
                }
                return nil
            }
            if node.Left != nil {
                q = append(q, node.Left)
            }
            if node.Right != nil {
                q = append(q, node.Right)
            }
        }
    }
    return nil
}
